File: src/main/services/integrations.ts
import net from "node:net";
import type {
  AsnDetails,
  GeoDetails,
  IntegrationSettings,
  PeeringDbDetails,
  ProviderId,
  ProviderStatus
} from "@common/ipc";
import { getLogger } from "./logger";
import { getSettingsStore } from "./persistence";

const log = getLogger();

type ProviderResult<T> = {
  status: ProviderStatus;
  value?: T;
};

type ProviderExecutor<T> = () => Promise<T | undefined>;

const providerQueues = new Map<ProviderId, Promise<void>>();
const lastRequestTimes = new Map<ProviderId, number>();

// Cache for PeeringDB results by ASN to avoid redundant queries
const peeringDbCache = new Map<number, { data: PeeringDbDetails | undefined; timestamp: number }>();
const PEERINGDB_CACHE_TTL = 1000 * 60 * 60 * 24; // 24 hours

// Rate limit delays in milliseconds per provider
const PROVIDER_RATE_LIMITS: Partial<Record<ProviderId, number>> = {
  "peeringdb": 5000,      // 5 seconds between PeeringDB requests (VERY strict to avoid 429)
  "team-cymru": 150,      // 150ms between Team Cymru requests
  "rdap": 250,            // 250ms between RDAP requests
  "ripe-stat": 250        // 250ms between RIPE Stat requests
};

// Track consecutive 429 errors per provider for exponential backoff
const rateLimitErrors = new Map<ProviderId, number>();
const MAX_BACKOFF_MS = 30000; // Maximum 30 second backoff

// Timeout values in milliseconds per provider
const PROVIDER_TIMEOUTS: Partial<Record<ProviderId, number>> = {
  "peeringdb": 10000,     // 10 seconds for PeeringDB
  "team-cymru": 8000,     // 8 seconds for Team Cymru (socket-based)
  "rdap": 10000,          // 10 seconds for RDAP
  "ripe-stat": 10000      // 10 seconds for RIPE Stat
};

async function enforceRateLimit(provider: ProviderId): Promise<void> {
  const minDelay = PROVIDER_RATE_LIMITS[provider];
  if (!minDelay) {
    return;
  }

  // Apply exponential backoff if we've had recent 429 errors
  const errorCount = rateLimitErrors.get(provider) ?? 0;
  const backoffMultiplier = Math.pow(2, errorCount); // 1x, 2x, 4x, 8x, etc.
  const backoffDelay = Math.min(minDelay * backoffMultiplier, MAX_BACKOFF_MS);

  const lastRequest = lastRequestTimes.get(provider);
  if (lastRequest) {
    const elapsed = Date.now() - lastRequest;
    const remaining = backoffDelay - elapsed;
    if (remaining > 0) {
      if (errorCount > 0) {
        log.warn(`[rate-limit] Exponential backoff (${errorCount} errors): delaying ${provider} request by ${remaining}ms`);
      } else {
        log.info(`[rate-limit] Delaying ${provider} request by ${remaining}ms to respect rate limit`);
      }
      await new Promise(resolve => setTimeout(resolve, remaining));
    }
  }
  lastRequestTimes.set(provider, Date.now());
}

async function withProviderQueue<T>(provider: ProviderId, fn: () => Promise<T>): Promise<T> {
  const previous = providerQueues.get(provider) ?? Promise.resolve();
  let release: (() => void) | undefined;

  const next = new Promise<void>((resolve) => {
    release = resolve;
  });

  providerQueues.set(
    provider,
    previous
      .catch(() => undefined)
      .then(() => next)
      .catch(() => undefined)
  );

  await previous.catch(() => undefined);

  try {
    // Enforce rate limiting before making the request
    await enforceRateLimit(provider);
    return await fn();
  } finally {
    release?.();
  }
}

async function executeProvider<T>(
  provider: ProviderId,
  enabled: boolean,
  executor: ProviderExecutor<T>,
  ipAddress: string,
  skipMessage?: string
): Promise<ProviderResult<T | undefined>> {
  if (!enabled) {
    const message = `[enrich] ${provider} skipped for ${ipAddress} — disabled in settings.`;
    log.info(message);
    return {
      status: { provider, status: "skipped", message: skipMessage ?? "Disabled in settings." }
    };
  }

  try {
    const startMessage = `[enrich] ${provider} lookup starting for ${ipAddress}`;
    log.info(startMessage);
    const value = await withProviderQueue(provider, () => executor());

    // Success - reset any rate limit error counter
    rateLimitErrors.delete(provider);

    if (value == null) {
      // No data found is normal, not an error - don't log or show as error
      log.info(`[enrich] ${provider} lookup finished for ${ipAddress} (no data found).`);
      return {
        status: { provider, status: "success", message: "No data found for this IP." }
      };
    }

    log.info(`[enrich] ${provider} lookup finished for ${ipAddress} (success).`);
    return {
      status: { provider, status: "success", details: value },
      value
    };
  } catch (error) {
    const message =
      error instanceof Error ? error.message : `Unexpected ${provider} error during lookup.`;

    // Check if it's a rate limit error
    const isRateLimitError = error instanceof Error && (
      error.message.includes("HTTP 429") ||
      error.message.includes("Rate limited") ||
      error.message.includes("rate limit")
    );

    // If rate limit error, increment counter for exponential backoff
    if (isRateLimitError) {
      const currentCount = rateLimitErrors.get(provider) ?? 0;
      rateLimitErrors.set(provider, currentCount + 1);
      log.warn(`[rate-limit] ${provider} rate limit error count: ${currentCount + 1}`);
    }

    // Check if it's a network-related error
    const isNetworkError = error instanceof Error && (
      error.message.includes("ENOTFOUND") ||
      error.message.includes("ECONNREFUSED") ||
      error.message.includes("ETIMEDOUT") ||
      error.message.includes("timeout") ||
      error.message.includes("fetch failed") ||
      error.message.includes("network")
    );

    const errorType = isRateLimitError ? "RATE LIMIT" : (isNetworkError ? "NETWORK ERROR" : "ERROR");
    const errorMessage = `[enrich] ${provider} lookup failed for ${ipAddress} [${errorType}]: ${message}`;
    log.warn(errorMessage);

    return {
      status: {
        provider,
        status: "error",
        message: isNetworkError ? `Network error: ${message}` : message
      }
    };
  }
}

interface TeamCymruResult {
  asn?: number;
  name?: string;
  prefix?: string;
  country?: string;
  registry?: string;
}

async function queryTeamCymru(ipAddress: string): Promise<TeamCymruResult | undefined> {
  return new Promise<TeamCymruResult | undefined>((resolve, reject) => {
    const socket = net.createConnection({ host: "whois.cymru.com", port: 43 });
    let payload = "";

    const timeout = PROVIDER_TIMEOUTS["team-cymru"] ?? 8000;
    socket.setTimeout(timeout, () => {
      socket.destroy(new Error("Team Cymru lookup timed out."));
    });

    socket.once("error", (error) => {
      reject(error);
    });

    socket.on("data", (chunk) => {
      payload += chunk.toString("utf8");
    });

    socket.once("end", () => {
      const lines = payload
        .split(/\r?\n/g)
        .map((line) => line.trim())
        .filter(Boolean);

      // Header is at index 0, first data row at index 1
      const dataRow = lines.find((line) => /^\d+\s*\|/.test(line));
      if (!dataRow) {
        resolve(undefined);
        return;
      }

      const parts = dataRow.split("|").map((value) => value.trim());
      if (parts.length < 7) {
        resolve(undefined);
        return;
      }

      const [asnStr, , prefix, country, registry, , name] = parts;
      const asn = Number.parseInt(asnStr, 10);
      resolve({
        asn: Number.isNaN(asn) ? undefined : asn,
        name,
        prefix,
        country: country || undefined,
        registry: registry || undefined
      });
    });

    socket.once("connect", () => {
      socket.write("begin\n");
      socket.write("verbose\n");
      socket.write(`${ipAddress}\n`);
      socket.end("end\n");
    });
  });
}

interface RdapResult {
  name?: string;
  handle?: string;
  country?: string;
  remarks?: string[];
}

async function queryRdap(ipAddress: string, baseUrl: string): Promise<RdapResult | undefined> {
  const url = `${baseUrl.replace(/\/$/, "")}/${encodeURIComponent(ipAddress)}`;
  const timeout = PROVIDER_TIMEOUTS["rdap"] ?? 10000;
  const response = await fetch(url, {
    headers: {
      "User-Agent": "VisTracer",
      Accept: "application/rdap+json, application/json"
    },
    signal: AbortSignal.timeout(timeout)
  });

  if (!response.ok) {
    if (response.status === 404) {
      return undefined;
    }

    if (response.status === 429) {
      throw new Error(`Rate limited by RDAP service (HTTP 429). Please wait before retrying.`);
    }

    throw new Error(`HTTP ${response.status} from RDAP service.`);
  }

  const data = (await response.json()) as {
    name?: string;
    handle?: string;
    country?: string;
    remarks?: { description?: string[] }[];
  };

  return {
    name: data.name,
    handle: data.handle,
    country: data.country,
    remarks: data.remarks?.flatMap((remark) => remark.description ?? []) ?? []
  };
}

interface RipeStatResult {
  asn?: number;
  holder?: string;
  prefix?: string;
  country?: string;
}

async function queryRipeStat(ipAddress: string, sourceApp: string): Promise<RipeStatResult | undefined> {
  const params = new URLSearchParams({
    resource: ipAddress,
    sourceapp: sourceApp
  });

  const timeout = PROVIDER_TIMEOUTS["ripe-stat"] ?? 10000;
  const response = await fetch(
    `https://stat.ripe.net/data/prefix-overview/data.json?${params.toString()}`,
    {
      headers: {
        "User-Agent": "VisTracer"
      },
      signal: AbortSignal.timeout(timeout)
    }
  );

  if (!response.ok) {
    if (response.status === 404) {
      return undefined;
    }

    if (response.status === 429) {
      throw new Error(`Rate limited by RIPE Stat (HTTP 429). Please wait before retrying.`);
    }

    throw new Error(`HTTP ${response.status} from RIPE Stat.`);
  }

  const payload = (await response.json()) as {
    data?: {
      asns?: { asn?: number; holder?: string; country?: string }[];
      prefix?: string;
    };
  };

  const primaryAsn = payload.data?.asns?.[0];

  return {
    asn: primaryAsn?.asn,
    holder: primaryAsn?.holder,
    country: primaryAsn?.country,
    prefix: payload.data?.prefix
  };
}

interface PeeringDbApiResponse {
  data?: Array<{
    id?: number;
    name?: string;
    aka?: string;
    website?: string;
    city?: string;
    country?: string;
    ix_count?: number;
  }>;
}

async function queryPeeringDb(asn: number, apiKey?: string): Promise<PeeringDbDetails | undefined> {
  // Check cache first to avoid redundant queries for the same ASN
  const cached = peeringDbCache.get(asn);
  if (cached && (Date.now() - cached.timestamp) < PEERINGDB_CACHE_TTL) {
    log.info(`[peeringdb] Cache hit for ASN ${asn}`);
    return cached.data;
  }

  const headers: Record<string, string> = {
    "User-Agent": "VisTracer"
  };

  if (apiKey) {
    headers["X-Api-Key"] = apiKey;
  }

  const timeout = PROVIDER_TIMEOUTS["peeringdb"] ?? 10000;
  const response = await fetch(`https://www.peeringdb.com/api/net?asn=${asn}`, {
    headers,
    signal: AbortSignal.timeout(timeout)
  });

  if (!response.ok) {
    if (response.status === 404) {
      // Cache the 404 result to avoid repeated queries
      peeringDbCache.set(asn, { data: undefined, timestamp: Date.now() });
      return undefined;
    }

    if (response.status === 429) {
      const retryAfter = response.headers.get("Retry-After");
      const waitTime = retryAfter ? `${retryAfter} seconds` : "later";
      throw new Error(`Rate limited by PeeringDB (HTTP 429). Retry after ${waitTime}. Consider using an API key for higher limits.`);
    }

    throw new Error(`HTTP ${response.status} from PeeringDB.`);
  }

  const payload = (await response.json()) as PeeringDbApiResponse;
  const entry = payload.data?.[0];

  if (!entry) {
    // Cache the empty result
    peeringDbCache.set(asn, { data: undefined, timestamp: Date.now() });
    return undefined;
  }

  const result = {
    id: entry.id,
    name: entry.name,
    aka: entry.aka,
    website: entry.website,
    city: entry.city,
    country: entry.country,
    ixCount: entry.ix_count
  };

  // Cache the successful result
  peeringDbCache.set(asn, { data: result, timestamp: Date.now() });
  return result;
}

export interface EnrichmentSeed {
  geo?: GeoDetails;
  asn?: AsnDetails;
}

export interface EnrichmentResult {
  geo?: GeoDetails;
  asn?: AsnDetails;
  providerStatuses: ProviderStatus[];
  peeringDb?: PeeringDbDetails;
}

function mergeAsn(base: AsnDetails | undefined, addition: Partial<AsnDetails>): AsnDetails {
  return {
    ...base,
    ...addition,
    network: addition.network ?? base?.network
  };
}

export async function enrichWithExternalProviders(
  ipAddress: string,
  seed: EnrichmentSeed
): Promise<EnrichmentResult> {
  const rawSettings = getSettingsStore().get("integrations") as IntegrationSettings | undefined;
  const settings: IntegrationSettings = {
    teamCymru: {
      enabled: rawSettings?.teamCymru?.enabled ?? true
    },
    rdap: {
      enabled: rawSettings?.rdap?.enabled ?? true,
      baseUrl: rawSettings?.rdap?.baseUrl ?? "https://rdap.org/ip"
    },
    ripeStat: {
      enabled: rawSettings?.ripeStat?.enabled ?? true,
      sourceApp: rawSettings?.ripeStat?.sourceApp ?? "VisTracer"
    },
    peeringDb: {
      enabled: rawSettings?.peeringDb?.enabled ?? false,
      apiKey: rawSettings?.peeringDb?.apiKey
    }
  };
  const providerStatuses: ProviderStatus[] = [];

  const settingsMessage = `[enrich] Provider toggles for ${ipAddress}: ` +
    `teamCymru=${settings.teamCymru.enabled}, rdap=${settings.rdap.enabled}, ` +
    `ripeStat=${settings.ripeStat.enabled}, peeringDb=${settings.peeringDb.enabled}`;
  log.info(settingsMessage);

  const introMessage = `[enrich] Starting enrichment pipeline for ${ipAddress}`;
  log.info(introMessage);

  if (seed.geo || seed.asn) {
    const seedMessage = `[enrich] MaxMind seed data for ${ipAddress}: hasGeo=${seed.geo != null}, hasAsn=${seed.asn != null}`;
    log.info(seedMessage);
    providerStatuses.push({
      provider: "maxmind",
      status: "success",
      message: "Resolved locally via GeoLite2."
    });
  } else {
    const noSeedMessage = `[enrich] No MaxMind seed data for ${ipAddress}, will attempt external providers only`;
    log.info(noSeedMessage);
    providerStatuses.push({
      provider: "maxmind",
      status: "error",
      message: "No GeoLite2 match found."
    });
  }

  let currentGeo = seed.geo;
  let currentAsn = seed.asn;
  let peeringDbDetails: PeeringDbDetails | undefined;

  const cymruResult = await executeProvider(
    "team-cymru",
    settings.teamCymru.enabled,
    () => queryTeamCymru(ipAddress),
    ipAddress
  );
  providerStatuses.push(cymruResult.status);

  if (cymruResult.value) {
    const beforeAsn = currentAsn?.asn;
    currentAsn = mergeAsn(currentAsn, {
      asn: currentAsn?.asn ?? cymruResult.value.asn,
      name: currentAsn?.name ?? cymruResult.value.name,
      network: currentAsn?.network ?? cymruResult.value.prefix,
      country: currentAsn?.country ?? cymruResult.value.country,
      registry: currentAsn?.registry ?? cymruResult.value.registry
    });
    const mergeMessage = `[enrich] Team Cymru data merged for ${ipAddress}: ASN=${currentAsn?.asn ?? 'none'} (was ${beforeAsn ?? 'none'})`;
    log.info(mergeMessage);
  }

  const rdapBaseUrl = settings.rdap.baseUrl?.trim() || "https://rdap.org/ip";
  const rdapResult = await executeProvider(
    "rdap",
    settings.rdap.enabled,
    () => queryRdap(ipAddress, rdapBaseUrl),
    ipAddress
  );
  providerStatuses.push(rdapResult.status);

  if (rdapResult.value) {
    let merged = false;
    if (!currentAsn?.name && rdapResult.value.name) {
      currentAsn = mergeAsn(currentAsn, { name: rdapResult.value.name });
      merged = true;
    }

    if (rdapResult.value.country && currentGeo?.country == null) {
      const baseline = currentGeo ?? seed.geo;
      currentGeo = {
        latitude: baseline?.latitude ?? 0,
        longitude: baseline?.longitude ?? 0,
        city: baseline?.city,
        country: rdapResult.value.country,
        isoCode: baseline?.isoCode,
        confidence: baseline?.confidence
      };
      merged = true;
    }

    if (merged) {
      const mergeMessage = `[enrich] RDAP data merged for ${ipAddress}`;
      log.info(mergeMessage);
    }
  }

  const ripeStatResult = await executeProvider(
    "ripe-stat",
    settings.ripeStat.enabled,
    () => queryRipeStat(ipAddress, settings.ripeStat.sourceApp || "VisTracer"),
    ipAddress
  );
  providerStatuses.push(ripeStatResult.status);

  if (ripeStatResult.value) {
    const beforeAsn = currentAsn?.asn;
    currentAsn = mergeAsn(currentAsn, {
      asn: currentAsn?.asn ?? ripeStatResult.value.asn,
      name: currentAsn?.name ?? ripeStatResult.value.holder,
      network: currentAsn?.network ?? ripeStatResult.value.prefix,
      country: currentAsn?.country ?? ripeStatResult.value.country
    });
    const mergeMessage = `[enrich] RIPE Stat data merged for ${ipAddress}: ASN=${currentAsn?.asn ?? 'none'} (was ${beforeAsn ?? 'none'})`;
    log.info(mergeMessage);
  }

  if (currentAsn?.asn) {
    const peeringDbResult = await executeProvider(
      "peeringdb",
      settings.peeringDb.enabled,
      () => queryPeeringDb(currentAsn!.asn!, settings.peeringDb.apiKey),
      ipAddress
    );
    providerStatuses.push(peeringDbResult.status);

    if (peeringDbResult.value) {
      peeringDbDetails = peeringDbResult.value;
      currentAsn = mergeAsn(currentAsn, {
        name: currentAsn.name ?? peeringDbResult.value.name,
        country: currentAsn.country ?? peeringDbResult.value.country
      });
      const mergeMessage = `[enrich] PeeringDB data merged for ${ipAddress}: ${peeringDbResult.value.name ?? 'unnamed'}`;
      log.info(mergeMessage);
    }
  } else {
    providerStatuses.push({
      provider: "peeringdb",
      status: "skipped",
      message: "ASN unknown, skipping PeeringDB lookup."
    });
  }

  const completeMessage = `[enrich] Completed enrichment pipeline for ${ipAddress}`;
  log.info(completeMessage);

  // Log summary of provider results
  const summary = providerStatuses.map(ps => `${ps.provider}:${ps.status}`).join(", ");
  const summaryMessage = `[enrich] Provider summary for ${ipAddress}: ${summary}`;
  log.info(summaryMessage);

  return {
    geo: currentGeo,
    asn: currentAsn,
    providerStatuses,
    peeringDb: peeringDbDetails
  };
}

File: src-tauri/src/commands/integrations.rs
use crate::commands::persistence::{IntegrationSettings, SharedStore};
use crate::types::*;
use std::io::{BufRead, BufReader, Write};
use std::net::TcpStream;
use std::time::Duration;

#[derive(Debug, Clone)]
pub struct EnrichmentResult {
    pub geo: Option<GeoDetails>,
    pub asn: Option<AsnDetails>,
    pub provider_statuses: Vec<ProviderStatus>,
    pub peering_db: Option<PeeringDbDetails>,
}

fn make_provider_status(provider: &str, status: &str, message: &str) -> ProviderStatus {
    ProviderStatus {
        provider: provider.to_string(),
        status: status.to_string(),
        message: Some(message.to_string()),
        details: None,
    }
}

#[derive(Debug)]
struct TeamCymruResult {
    asn: Option<u32>,
    name: Option<String>,
    prefix: Option<String>,
    country: Option<String>,
    registry: Option<String>,
}

async fn query_team_cymru(ip: &str) -> Result<Option<TeamCymruResult>, String> {
    let ip_str = ip.to_string();
    tokio::task::spawn_blocking(move || {
        let mut stream = TcpStream::connect_timeout(
            &"whois.cymru.com:43".parse().map_err(|e| format!("{}", e))?,
            Duration::from_secs(8),
        )
        .map_err(|e| format!("Team Cymru connection failed: {}", e))?;

        stream
            .set_read_timeout(Some(Duration::from_secs(8)))
            .ok();

        stream
            .write_all(format!("begin\nverbose\n{}\nend\n", ip_str).as_bytes())
            .map_err(|e| format!("Write failed: {}", e))?;

        let reader = BufReader::new(stream);
        let mut data_row: Option<String> = None;

        for line in reader.lines() {
            let line = line.map_err(|e| format!("Read failed: {}", e))?;
            let trimmed = line.trim().to_string();
            if trimmed.is_empty() {
                continue;
            }
            // Data rows start with a digit followed by pipe
            if trimmed.chars().next().map_or(false, |c| c.is_ascii_digit())
                && trimmed.contains('|')
            {
                data_row = Some(trimmed);
                break;
            }
        }

        match data_row {
            None => Ok(None),
            Some(row) => {
                let parts: Vec<&str> = row.split('|').map(|s| s.trim()).collect();
                if parts.len() < 7 {
                    return Ok(None);
                }
                let asn = parts[0].parse::<u32>().ok();
                let prefix = if parts[2].is_empty() {
                    None
                } else {
                    Some(parts[2].to_string())
                };
                let country = if parts[3].is_empty() {
                    None
                } else {
                    Some(parts[3].to_string())
                };
                let registry = if parts[4].is_empty() {
                    None
                } else {
                    Some(parts[4].to_string())
                };
                let name = if parts.len() > 6 && !parts[6].is_empty() {
                    Some(parts[6].to_string())
                } else {
                    None
                };

                Ok(Some(TeamCymruResult {
                    asn,
                    name,
                    prefix,
                    country,
                    registry,
                }))
            }
        }
    })
    .await
    .map_err(|e| format!("Task join error: {}", e))?
}

#[derive(Debug)]
struct RdapResult {
    name: Option<String>,
    country: Option<String>,
}

async fn query_rdap(ip: &str, base_url: &str) -> Result<Option<RdapResult>, String> {
    let url = format!("{}/{}", base_url.trim_end_matches('/'), ip);
    let client = reqwest::Client::new();
    let resp = client
        .get(&url)
        .header("User-Agent", "VisTracer")
        .header("Accept", "application/rdap+json, application/json")
        .timeout(Duration::from_secs(10))
        .send()
        .await
        .map_err(|e| format!("RDAP request failed: {}", e))?;

    if resp.status() == reqwest::StatusCode::NOT_FOUND {
        return Ok(None);
    }
    if resp.status() == reqwest::StatusCode::TOO_MANY_REQUESTS {
        return Err("Rate limited by RDAP service (HTTP 429).".to_string());
    }
    if !resp.status().is_success() {
        return Err(format!("HTTP {} from RDAP service.", resp.status()));
    }

    let data: serde_json::Value = resp
        .json()
        .await
        .map_err(|e| format!("RDAP JSON parse failed: {}", e))?;

    Ok(Some(RdapResult {
        name: data["name"].as_str().map(|s| s.to_string()),
        country: data["country"].as_str().map(|s| s.to_string()),
    }))
}

#[derive(Debug)]
struct RipeStatResult {
    asn: Option<u32>,
    holder: Option<String>,
    prefix: Option<String>,
    country: Option<String>,
}

async fn query_ripe_stat(ip: &str, source_app: &str) -> Result<Option<RipeStatResult>, String> {
    let url = format!(
        "https://stat.ripe.net/data/prefix-overview/data.json?resource={}&sourceapp={}",
        ip, source_app
    );
    let client = reqwest::Client::new();
    let resp = client
        .get(&url)
        .header("User-Agent", "VisTracer")
        .timeout(Duration::from_secs(10))
        .send()
        .await
        .map_err(|e| format!("RIPE Stat request failed: {}", e))?;

    if resp.status() == reqwest::StatusCode::NOT_FOUND {
        return Ok(None);
    }
    if resp.status() == reqwest::StatusCode::TOO_MANY_REQUESTS {
        return Err("Rate limited by RIPE Stat (HTTP 429).".to_string());
    }
    if !resp.status().is_success() {
        return Err(format!("HTTP {} from RIPE Stat.", resp.status()));
    }

    let data: serde_json::Value = resp
        .json()
        .await
        .map_err(|e| format!("RIPE Stat JSON parse failed: {}", e))?;

    let primary_asn = &data["data"]["asns"][0];

    Ok(Some(RipeStatResult {
        asn: primary_asn["asn"].as_u64().map(|v| v as u32),
        holder: primary_asn["holder"].as_str().map(|s| s.to_string()),
        country: primary_asn["country"].as_str().map(|s| s.to_string()),
        prefix: data["data"]["prefix"].as_str().map(|s| s.to_string()),
    }))
}

async fn query_peeringdb(
    asn: u32,
    api_key: Option<&str>,
) -> Result<Option<PeeringDbDetails>, String> {
    let url = format!("https://www.peeringdb.com/api/net?asn={}", asn);
    let client = reqwest::Client::new();
    let mut req = client
        .get(&url)
        .header("User-Agent", "VisTracer")
        .timeout(Duration::from_secs(10));

    if let Some(key) = api_key {
        req = req.header("X-Api-Key", key);
    }

    let resp = req
        .send()
        .await
        .map_err(|e| format!("PeeringDB request failed: {}", e))?;

    if resp.status() == reqwest::StatusCode::NOT_FOUND {
        return Ok(None);
    }
    if resp.status() == reqwest::StatusCode::TOO_MANY_REQUESTS {
        return Err("Rate limited by PeeringDB (HTTP 429).".to_string());
    }
    if !resp.status().is_success() {
        return Err(format!("HTTP {} from PeeringDB.", resp.status()));
    }

    let data: serde_json::Value = resp
        .json()
        .await
        .map_err(|e| format!("PeeringDB JSON parse failed: {}", e))?;

    let entry = &data["data"][0];
    if entry.is_null() {
        return Ok(None);
    }

    Ok(Some(PeeringDbDetails {
        id: entry["id"].as_u64(),
        name: entry["name"].as_str().map(|s| s.to_string()),
        aka: entry["aka"].as_str().map(|s| s.to_string()),
        website: entry["website"].as_str().map(|s| s.to_string()),
        city: entry["city"].as_str().map(|s| s.to_string()),
        country: entry["country"].as_str().map(|s| s.to_string()),
        ix_count: entry["ix_count"].as_u64(),
    }))
}

fn merge_asn(base: Option<&AsnDetails>, addition: &AsnDetails) -> AsnDetails {
    AsnDetails {
        asn: addition.asn.or_else(|| base.and_then(|b| b.asn)),
        name: addition
            .name
            .clone()
            .or_else(|| base.and_then(|b| b.name.clone())),
        network: addition
            .network
            .clone()
            .or_else(|| base.and_then(|b| b.network.clone())),
        country: addition
            .country
            .clone()
            .or_else(|| base.and_then(|b| b.country.clone())),
        registry: addition
            .registry
            .clone()
            .or_else(|| base.and_then(|b| b.registry.clone())),
    }
}

pub async fn enrich_with_external_providers(
    ip: &str,
    seed_geo: Option<&GeoDetails>,
    seed_asn: Option<&AsnDetails>,
    store: &SharedStore,
) -> EnrichmentResult {
    let settings = {
        let guard = store.lock().unwrap();
        guard.integrations.clone()
    };

    let mut statuses: Vec<ProviderStatus> = Vec::new();
    let mut current_geo = seed_geo.cloned();
    let mut current_asn = seed_asn.cloned();
    let mut peering_db_details: Option<PeeringDbDetails> = None;

    // MaxMind status
    if seed_geo.is_some() || seed_asn.is_some() {
        statuses.push(make_provider_status(
            "maxmind",
            "success",
            "Resolved locally via GeoLite2.",
        ));
    } else {
        statuses.push(make_provider_status(
            "maxmind",
            "error",
            "No GeoLite2 match found.",
        ));
    }

    // Team Cymru
    if settings.team_cymru.enabled {
        match query_team_cymru(ip).await {
            Ok(Some(result)) => {
                let addition = AsnDetails {
                    asn: result.asn,
                    name: result.name,
                    network: result.prefix,
                    country: result.country,
                    registry: result.registry,
                };
                current_asn = Some(merge_asn(current_asn.as_ref(), &addition));
                statuses.push(make_provider_status("team-cymru", "success", "Lookup complete."));
            }
            Ok(None) => {
                statuses.push(make_provider_status(
                    "team-cymru",
                    "success",
                    "No data found for this IP.",
                ));
            }
            Err(e) => {
                statuses.push(make_provider_status("team-cymru", "error", &e));
            }
        }
    } else {
        statuses.push(make_provider_status(
            "team-cymru",
            "skipped",
            "Disabled in settings.",
        ));
    }

    // RDAP
    if settings.rdap.enabled {
        let base_url = settings
            .rdap
            .base_url
            .as_deref()
            .unwrap_or("https://rdap.org/ip");
        match query_rdap(ip, base_url).await {
            Ok(Some(result)) => {
                if current_asn.as_ref().and_then(|a| a.name.as_ref()).is_none() {
                    if let Some(name) = result.name {
                        let addition = AsnDetails {
                            asn: None,
                            name: Some(name),
                            network: None,
                            country: None,
                            registry: None,
                        };
                        current_asn = Some(merge_asn(current_asn.as_ref(), &addition));
                    }
                }
                if current_geo.as_ref().and_then(|g| g.country.as_ref()).is_none() {
                    if let Some(country) = result.country {
                        if let Some(ref mut geo) = current_geo {
                            geo.country = Some(country);
                        }
                    }
                }
                statuses.push(make_provider_status("rdap", "success", "Lookup complete."));
            }
            Ok(None) => {
                statuses.push(make_provider_status(
                    "rdap",
                    "success",
                    "No data found for this IP.",
                ));
            }
            Err(e) => {
                statuses.push(make_provider_status("rdap", "error", &e));
            }
        }
    } else {
        statuses.push(make_provider_status(
            "rdap",
            "skipped",
            "Disabled in settings.",
        ));
    }

    // RIPE Stat
    if settings.ripe_stat.enabled {
        let source_app = &settings.ripe_stat.source_app;
        match query_ripe_stat(ip, source_app).await {
            Ok(Some(result)) => {
                let addition = AsnDetails {
                    asn: result.asn,
                    name: result.holder,
                    network: result.prefix,
                    country: result.country,
                    registry: None,
                };
                current_asn = Some(merge_asn(current_asn.as_ref(), &addition));
                statuses.push(make_provider_status("ripe-stat", "success", "Lookup complete."));
            }
            Ok(None) => {
                statuses.push(make_provider_status(
                    "ripe-stat",
                    "success",
                    "No data found for this IP.",
                ));
            }
            Err(e) => {
                statuses.push(make_provider_status("ripe-stat", "error", &e));
            }
        }
    } else {
        statuses.push(make_provider_status(
            "ripe-stat",
            "skipped",
            "Disabled in settings.",
        ));
    }

    // PeeringDB
    if let Some(asn_num) = current_asn.as_ref().and_then(|a| a.asn) {
        if settings.peering_db.enabled {
            let api_key = settings.peering_db.api_key.as_deref();
            match query_peeringdb(asn_num, api_key).await {
                Ok(Some(result)) => {
                    // Merge PeeringDB data into ASN
                    if current_asn.as_ref().and_then(|a| a.name.as_ref()).is_none() {
                        if let Some(ref name) = result.name {
                            let addition = AsnDetails {
                                asn: None,
                                name: Some(name.clone()),
                                network: None,
                                country: result.country.clone(),
                                registry: None,
                            };
                            current_asn = Some(merge_asn(current_asn.as_ref(), &addition));
                        }
                    }
                    peering_db_details = Some(result);
                    statuses.push(make_provider_status(
                        "peeringdb",
                        "success",
                        "Lookup complete.",
                    ));
                }
                Ok(None) => {
                    statuses.push(make_provider_status(
                        "peeringdb",
                        "success",
                        "No data found for this ASN.",
                    ));
                }
                Err(e) => {
                    statuses.push(make_provider_status("peeringdb", "error", &e));
                }
            }
        } else {
            statuses.push(make_provider_status(
                "peeringdb",
                "skipped",
                "Disabled in settings.",
            ));
        }
    } else {
        statuses.push(make_provider_status(
            "peeringdb",
            "skipped",
            "ASN unknown, skipping PeeringDB lookup.",
        ));
    }

    EnrichmentResult {
        geo: current_geo,
        asn: current_asn,
        provider_statuses: statuses,
        peering_db: peering_db_details,
    }
}

pub fn get_integration_settings(store: &SharedStore) -> IntegrationSettings {
    let guard = store.lock().unwrap();
    guard.integrations.clone()
}

pub fn set_integration_settings(store: &SharedStore, settings: IntegrationSettings) {
    let mut guard = store.lock().unwrap();
    guard.integrations = settings;
}
